src/transport/SecureMessageCodec.cpp

```diff
@@ -0,0 +1,139 @@
+ */
+
+/**
+ *    @file
+ *      This file defines functions for encoding and decoding CHIP messages.
+ *      The encoded messages contain CHIP packet header, encrypted payload
+ *      header, encrypted payload and message authentication code, as per
+ *      CHIP specifications.
+ *
+ */
+
+#include <support/CodeUtils.h>
+#include <support/ReturnMacros.h>
+#include <support/SafeInt.h>
+#include <transport/SecureMessageCodec.h>
+
+// Maximum length of application data that can be encrypted as one block.
+// The limit is derived from IPv6 MTU (1280 bytes) - expected header overheads.
+// This limit would need additional reviews once we have formalized Secure Transport header.
+//
+// TODO: this should be checked within the transport message sending instead of the session management layer.
+static const size_t kMax_SecureSDU_Length = 1024;
+
+namespace chip {
+
+using System::PacketBuffer;
+using System::PacketBufferHandle;
+
+namespace SecureMessageCodec {
+
+CHIP_ERROR Encode(NodeId localNodeId, Transport::PeerConnectionState * state, PayloadHeader & payloadHeader,
+                  PacketHeader & packetHeader, System::PacketBufferHandle & msgBuf)
+{
+    VerifyOrReturnError(!msgBuf.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(!msgBuf->HasChainedBuffer(), CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    VerifyOrReturnError(msgBuf->TotalLength() < kMax_SecureSDU_Length, CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    uint32_t msgId = state->GetSendMessageIndex();
+
+    static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                  "Addition to generate payloadLength might overflow");
+
+    uint16_t headerSize = payloadHeader.EncodeSizeBytes();
+
+    // Make sure it's big enough to add two 16-bit ints without overflowing.
+    uint32_t payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+    VerifyOrReturnError(CanCastTo<uint16_t>(payloadLength), CHIP_ERROR_NO_MEMORY);
+
+    packetHeader
+        .SetSourceNodeId(localNodeId)                 //
+        .SetDestinationNodeId(state->GetPeerNodeId()) //
+        .SetMessageId(msgId)                          //
+        .SetEncryptionKeyID(state->GetLocalKeyID())   //
+        .SetPayloadLength(static_cast<uint16_t>(payloadLength));
+    packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
+
+    VerifyOrReturnError(msgBuf->EnsureReservedSize(headerSize), CHIP_ERROR_NO_MEMORY);
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
+    uint8_t * data    = msgBuf->Start();
+    uint16_t totalLen = msgBuf->TotalLength();
+
+    uint16_t actualEncodedHeaderSize;
+    ReturnErrorOnFailure(payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize));
+
+    MessageAuthenticationCode mac;
+    ReturnErrorOnFailure(state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac));
+
+    uint16_t taglen = 0;
+    ReturnErrorOnFailure(mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen));
```

Comment:
I hit this yesterday in the master version — there's no check that the packet buffer has room for this (e.g. `msgBuf->AvailableDataLength() >= kMaxTagLen`).

Some tests do not in fact request enough space, and currently work only because all current builds actually use fixed-size buffer pools. I have a near-future PR that includes changes to those tests, but optionally you can grab test changes from https://github.com/kpschoedel/connectedhomeip/tree/x4395-encrypt-tests

Reply:
Thanks @kpschoedel, that's a good catch. The `mac.Encode` checks for the size before filling in the value, but the size was hardcoded to `kMaxTagLen`. I've updated the code to use `msgBuf->AvailableDataLength()` as the input size. This should help catch scenarios where the packet buffer is not large enough.